cxx/util_distribution_variant.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#include "util_distribution_variant.hh"
+
+#include <cassert>
+#include <sstream>
+
+#include "distributions/beta_bernoulli.hh"
+#include "distributions/bigram.hh"
+#include "distributions/crp.hh"
+#include "distributions/dirichlet_categorical.hh"
+#include "distributions/normal.hh"
+#include "domain.hh"
+#include "relation.hh"
+
+ObservationVariant observation_string_to_value(
+    const std::string& value_str, const DistributionEnum& distribution) {
+  switch (distribution) {
+    case DistributionEnum::normal:
+      return std::stod(value_str);
+    case DistributionEnum::bernoulli:
+      return std::stod(value_str);
+    case DistributionEnum::categorical:
+      return std::stoi(value_str);
+    case DistributionEnum::bigram:
+      return value_str;
+    default:
+      assert(false);
+  }
+}
+
+DistributionSpec parse_distribution_spec(const std::string& dist_str) {
+  DistributionEnum dist;
+  std::string prefix;
+  std::map<std::string, std::string> dist_args;
+  if (dist_str.starts_with("bernoulli")) {
+    prefix = "bernoulli";
+    dist = DistributionEnum::bernoulli;
+    assert(prefix.size() == dist_str.size());
+  } else if (dist_str.starts_with("bigram")) {
+    prefix = "bigram";
+    dist = DistributionEnum::bigram;
+    assert(prefix.size() == dist_str.size());
+  } else if (dist_str.starts_with("categorical")) {
+    prefix = "categorical";
+    dist = DistributionEnum::categorical;
+  } else if (dist_str.starts_with("normal")) {
+    prefix = "normal";
+    dist = DistributionEnum::normal;
+    assert(prefix.size() == dist_str.size());
+  } else {
+    assert(false);
+  }
+  std::string args_str = dist_str.substr(prefix.length());
+  if (args_str.empty()) {
+    return DistributionSpec{dist};
+  } else {
+    assert(args_str[0] == '(');
```

Comment:
This would be a lot cleaner if it used regular expressions (using Boost.Regex say).  For example, that way you could parse out the distribution name and then look it up in a map from names to DistributionEnum values and not have to do it through a chain of if's.

Reply:
I have not tested the following code, but it should be something like

#include <boost/regex.hpp>

boost::regex re{"^(\\w+)(?:\(((\w+)\s*=\s*(\w+),?)*\))?$"};
boost:smatch match;

if (boost::regex_search(dist_str, match, re)) {
  string dist_name = match[1];
  
  // First keyword will be in match[2][0] and its value in match[2][1] etc.
}